### ISSUE TITLE
Set Host header has no effect

### ISSUE THREAD

**Issue by zunkree** [containous/traefik#6500]
### Do you want to request a *feature* or report a *bug*?
Bug

### What did you do?
Set `Host` header using `Headers` middleware.

```yaml
  middlewares:
    example-headers:
      headers:
        customRequestHeaders:
          Host: "example.com"
```

### What did you expect to see?
Correct set `Host` in HTTP request object. Due to https://github.com/golang/go/issues/29865 `Host` field of `http.Request` should be set directly because of `Request.Write` ignore `Host` header:
```json
{
  "Method": "GET",
  "URL": {
    "Scheme": "",
    "Opaque": "",
    "User": null,
    "Host": "",
    "Path": "/status",
    "RawPath": "",
    "ForceQuery": false,
    "RawQuery": "format=json",
    "Fragment": ""
  },
  "Proto": "HTTP/1.1",
  "ProtoMajor": 1,
  "ProtoMinor": 1,
  "Header": {
    "Accept": [
      "*/*"
    ],
    "Host": [
      "example.com"
    ],
    "User-Agent": [
      "curl/7.58.0"
    ],
    "X-Forwarded-Host": [
      "127.0.0.1:8080"
    ],
    "X-Forwarded-Port": [
      "8080"
    ],
    "X-Forwarded-Proto": [
      "http"
    ],
    "X-Forwarded-Server": [
      "lb.example.com"
    ],
    "X-Real-Ip": [
      "127.0.0.1"
    ]
  },
  "ContentLength": 0,
  "TransferEncoding": null,
  "Host": "example.com",
  "Form": null,
  "PostForm": null,
  "MultipartForm": null,
  "Trailer": null,
  "RemoteAddr": "127.0.0.1:13094",
  "RequestURI": "/status?format=json",
  "TLS": null
}
```

### What did you see instead?
`Host` field of `http.Request` is not set correctly:
```json
{
  "Method": "GET",
  "URL": {
    "Scheme": "",
    "Opaque": "",
    "User": null,
    "Host": "",
    "Path": "/status",
    "RawPath": "",
    "ForceQuery": false,
    "RawQuery": "format=json",
    "Fragment": ""
  },
  "Proto": "HTTP/1.1",
  "ProtoMajor": 1,
  "ProtoMinor": 1,
  "Header": {
    "Accept": [
      "*/*"
    ],
    "Host": [
      "example.com"
    ],
    "User-Agent": [
      "curl/7.58.0"
    ],
    "X-Forwarded-Host": [
      "127.0.0.1:8080"
    ],
    "X-Forwarded-Port": [
      "8080"
    ],
    "X-Forwarded-Proto": [
      "http"
    ],
    "X-Forwarded-Server": [
      "lb.example.com"
    ],
    "X-Real-Ip": [
      "127.0.0.1"
    ]
  },
  "ContentLength": 0,
  "TransferEncoding": null,
  "Host": "127.0.0.1:8080",
  "Form": null,
  "PostForm": null,
  "MultipartForm": null,
  "Trailer": null,
  "RemoteAddr": "127.0.0.1:13094",
  "RequestURI": "/status?format=json",
  "TLS": null
}
```

### Output of `traefik version`:
```
$ docker run -ti traefik:v2.2.0-rc2 version
Version:      2.2.0-rc2
Codename:     chevrotin
Go version:   go1.14
Built:        2020-03-11T17:42:16Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

`traefik.yaml`:
```yaml
---
entryPoints:
  http:
    address: '0.0.0.0:8080'
  metrics:
    address: '0.0.0.0:9000'
metrics:
  prometheus:
    entryPoint: metrics
log:
  level: DEBUG
accessLog: {}
providers:
  file:
    watch: true
    filename: "/etc/traefik/example-lb.yml"
    debugLogGeneratedTemplate: true
```

`example-lb.yml`:
```yaml
---
http:
  routers:
    example:
      entryPoints:
      - http
      middlewares:
      - example-headers
      - error-ratio-cb
      service: example
      rule: "PathPrefix(`/`)"

  middlewares:
    example-headers:
      headers:
        customRequestHeaders:
          Host: "example.com"
    error-ratio-cb:
      circuitBreaker:
        expression: 'ResponseCodeRatio(500, 600, 0, 600) > 0.10'

  services:
    example:
      loadBalancer:
        passHostHeader: true
        servers:
        - url: "https://origin.example.com"
```

**Comment by ldez** [comment]
duplicate of #5966, #6000,

**Comment by zunkree** [comment]
@ldez do you have any plans to fix it? It is quite crucial for our use case and I can help you with PR. The only question here is the where is the best place to modify `http.Request.Host` field, in middleware itself or in `proxy` package. Or maybe there is another place to do that, what do you think?

**Comment by ldez** [comment]
do you have a concrete use case ?

**Comment by zunkree** [comment]
We also need to pass `Host` with specific value. That is set of dynamically spawning servers with bunch of virtual hosts on them and we need to separate requests using virtual hosts. In the future we plan to get rid of that approach but for now traefik here is the best option for load balancer because it provides great observability and performance so we decided to move to traefik as load balancer in front of our fleet and then make any changes with better understanding how it is affect our platform. Also, AWS API Gateway requires that: https://stackoverflow.com/questions/39591495/nginx-proxy-pass-to-aws-api-gateway

**Comment by ldez** [comment]
Have to try to set `passHostHeader` to false (the default value is true)

https://docs.traefik.io/v2.1/routing/services/#pass-host-header

**Comment by zunkree** [comment]
I've tried, has no effect unfortunatelly. I did some quick hack and looks like it works:
```diff
diff --git a/pkg/middlewares/headers/headers.go b/pkg/middlewares/headers/headers.go
index 4868c8cf..38663ea9 100644
--- a/pkg/middlewares/headers/headers.go
+++ b/pkg/middlewares/headers/headers.go
@@ -169,6 +169,9 @@ func (s *Header) modifyCustomRequestHeaders(req *http.Request) {
                        req.Header.Del(header)
                } else {
                        req.Header.Set(header, value)
+                       if header == "Host" {
+                               req.Host = value
+                       }
                }
        }
 }

```

**Comment by zunkree** [comment]
thank you very much guys!

**Comment by traefiker** [comment]
Closed by #6502.
